File: archive/t/typescript/longest-palindromic-substring.ts
function printUsage(): void {
  console.error(
    "Usage: please provide a string that contains at least one palindrome",
  );
  process.exit(1);
}

function preprocess(s: string): string {
  // e.g. "abba" -> "^#a#b#b#a#$"
  return "^#" + s.split("").join("#") + "#$";
}

function longestPalindromicSubstring(s: string): string | null {
  if (!s) return null;

  const T = preprocess(s);
  const n = T.length;
  const P = new Array<number>(n).fill(0);

  let C = 0; // center
  let R = 0; // right boundary

  for (let i = 1; i < n - 1; i++) {
    const mirror = 2 * C - i;

    if (i < R) {
      P[i] = Math.min(R - i, P[mirror]);
    }

    // expand around center i
    while (T[i + 1 + P[i]] === T[i - 1 - P[i]]) {
      P[i]++;
    }

    // update center
    if (i + P[i] > R) {
      C = i;
      R = i + P[i];
    }
  }

  // find max palindrome
  let maxLen = 0;
  let centerIndex = 0;

  for (let i = 1; i < n - 1; i++) {
    if (P[i] > maxLen) {
      maxLen = P[i];
      centerIndex = i;
    }
  }

  if (maxLen === 0) return null;

  const start = (centerIndex - maxLen) >> 1;
  return s.substring(start, start + maxLen);
}

function main(): void {
  const input = process.argv[2];

  if (!input) {
    printUsage();
  }

  const result = longestPalindromicSubstring(input);
  if (!result || result?.length < 2) {
    printUsage();
  }

  console.log(result);
}

main();

File: archive/v/visual-basic/longest-palindromic-substring.vb
Public Module LongestPalindromicSubstring

    Public Sub Main(args As String())

        If args.Length <> 1 Then
            Console.WriteLine("Usage: please provide a string that contains at least one palindrome")
            Return
        End If

        Dim input As String = args(0)

        If String.IsNullOrWhiteSpace(input) Then
            Console.WriteLine("Usage: please provide a string that contains at least one palindrome")
            Return
        End If

        Dim result = LongestPalindrome(input)
        If result.Length < 2 Then
            Console.WriteLine("Usage: please provide a string that contains at least one palindrome")
            Return
        End If

        Console.WriteLine(result)

    End Sub


    Public Function LongestPalindrome(s As String) As String

        Dim t As String = Preprocess(s)

        Dim n As Integer = t.Length
        Dim p(n - 1) As Integer

        Dim center As Integer = 0
        Dim right As Integer = 0

        Dim maxLen As Integer = 0
        Dim maxCenter As Integer = 0

        For i As Integer = 1 To n - 2

            Dim mirror As Integer = 2 * center - i

            If i < right Then
                p(i) = Math.Min(right - i, p(mirror))
            End If

            While t(i + p(i) + 1) = t(i - p(i) - 1)
                p(i) += 1
            End While

            If i + p(i) > right Then
                center = i
                right = i + p(i)
            End If

            If p(i) > maxLen Then
                maxLen = p(i)
                maxCenter = i
            End If

        Next

        Dim startIndex As Integer = (maxCenter - maxLen) \ 2
        Return s.Substring(startIndex, maxLen)

    End Function


    Private Function Preprocess(s As String) As String

        Dim result As String = "^#"

        For Each c As Char In s
            result &= c & "#"
        Next

        result &= "$"

        Return result

    End Function

End Module